File: Final/Class_ising.py
import numpy as np
import math
import matplotlib
import matplotlib.pyplot as plt

from functions import get_init_energy  #compute energies
from functions import energy_difference  #compute energy differences
#these are externally defined, 

#metropolis criterion
def metropolis(e,enew):  
	if(e>enew):
		return True
	else:
		if np.random.random()<math.exp((e-enew)):
			return True
		else:
			return False


N=50  #size of the lattice
nstep=500000	#steps in the simulation
printstep=5000	#steps between printing
nprint=int(nstep/printstep)
nrun=100

t=np.zeros(nprint)
avm=np.zeros(nprint)
ave=np.zeros(nprint)


np.random.seed(2)

J=0.1  #interaction strength (with kT=1)
b=0  #magnetic field.  This is h in class.  

pbc=True  #this flags whether to use PBCs

for run in range(nrun):

		
	spins=[[0 for _ in range(N)] for _ in range(N)]  #define the spins as all down
	m=0  #initial magnetization is all down
	for i in range(N):
		for j in range(N):
			if np.random.random()<.5:
				spins[i][j]=1  
				m+=1
			else:
				spins[i][j]=-1
				m-=1;			
	e=get_init_energy(spins,J,b,N,pbc)
	#compute initial energy


	for step in range(nstep):  #iterate over steps
		s1=np.random.randint(0,N)
		s2=np.random.randint(0,N)
		#randomly choose a spin to flip.  		
		
		
		enew=e+energy_difference(s1,s2,spins,J,b,N,pbc)
		#compute the energy
		if(metropolis(e,enew)):
		#with the metropolis criterion, choose to flip the spin
			spins[s1][s2]=-spins[s1][s2]
			if spins[s1][s2]>0:
				m+=2
			else:
				m-=2
			#update the magnetization due tot he flip
			e=enew
			#update the energy due to the flip
	
		if step%printstep==0:
			t[int(step/printstep)]=step
			#every printstep, we want to print data
			avm[int(step/printstep)]+=abs(m)/N/N
			ave[int(step/printstep)]+=e/J/N/N

s=np.transpose(spins)
for i in range(N):
	y=[j for j,e in enumerate(s[i]) if e==1]
	x=[i for _ in range(len(y))]
	plt.scatter(x,y,color='b')
ax1=plt.axes()
ax1.set_aspect('equal')
print(y)
print(x)
plt.show()



plt.plot(t/nrun,avm/nrun)
plt.show()
plt.plot(t/nrun,ave/nrun)
plt.show()

File: Final/class_ising.m
clear;
rng(2,'twister');

N=10;
nstep=500000;
printstep=5000;
nprint=floor(nstep/printstep)+1;
nrun=100;

t=zeros(nprint,1);
avm=zeros(nprint,1);
ave=zeros(nprint,1);

J=0.1;
b=0;
pbc=true;

for run=1:nrun
   spins=zeros(N,N);
   m=0;
   for i=1:N
        for j=1:N
            if(rand()<.5)
                spins(i,j)=1;
                m=m+1;
            else
                spins(i,j)=-1;
                m=m-1;
            end
        end
   end
   e=get_init_energy(spins,J,b,N,pbc);
   
   if(run==1)
        up=0;
        down=0;
    for i=1:N
        for j=1:N
            hold on
            if(spins(i,j)==1)
                plot(i,j,"ro")
                up=up+1;
            elseif(spins(i,j)==-1)
                plot(i,j,"bx")
                down=down+1;
            end
        end
    end
   legend(sprintf('down %g',down),sprintf('up %g',up));
   end
   for step=1:nstep
      s1=floor(N*rand())+1;
      s2=floor(N*rand())+1;
      enew=e+energy_difference(s1,s2,spins,J,b,N,pbc);
      if(metropolis(e,enew))
         spins(s1,s2)=-spins(s1,s2);
         if(spins(s1,s2)>0)
            m=m+2; 
         else
             m=m-2;
         end
      end
      e=enew;
      
      if(mod(step,printstep)==0)
          k=floor(step/printstep)+1;
          t(k)=step;
          avm(k)=avm(k)+abs(m)/N/N;
          ave(k)=ave(k)+e/J/N/N;
      end
   end
end
up=0;
down=0;
figure
for i=1:N
        for j=1:N
            hold on
            if(spins(i,j)==1)
                plot(i,j,"ro")
                up=up+1;
            elseif(spins(i,j)==-1)
                plot(i,j,"bx")
                down=down+1;
            end
        end
end
legend(sprintf('down %g',down),sprintf('up %g',up));

figure
plot(t,ave./nrun)
figure
plot(t,avm./nrun)


function x=metropolis(e,enew)
   if(e>enew)
       x=true;
   else
       if(rand()<exp(-(e-enew)))
           x=true;
       else
           x=false;
       end
   end
end


function e=get_init_energy(spins,J,b,N,pbc)
    e=0;
end


function de=energy_difference(s1,s2,spins,J,b,N,pbc)
    de=0;
end
